### PR TITLE
Related diversify publishers

### PR DESCRIPTION
**Pull request by eu9ene**
# Goal

- Add the diversification of publishers for after-article recommendations on the syndicated page.
- Update to v2 qdrant collection requires running the pipeline from https://github.com/Pocket/dl-metaflow-jobs/pull/166 first
- Make sure syndicated items are not shown on the right rail of the syndicated page

# Todo

- [x] merge the metaflow pipeline and run it to get the collection with new domains

# Links
- [Jira ticket](https://getpocket.atlassian.net/browse/TDP-295)

### CODE REVIEW

File: app/data_providers/dispatch.py [hunk 1]
```diff
@@ -51,25 +52,29 @@ async def after_save(self, resolved_id: int, count: int) -> List[CorpusRecommend
...
             # do not fallback for "Similar stores" after saving
             recs = []
-        return self._sample(recs, count)
+        return self._to_corpus_items(recs, count)
 
     @_empty_on_error
     async def after_article(self, resolved_id: int, count: int) -> List[CorpusRecommendationModel]:
         try:
-            recs = await self.item_recommender.related(resolved_id, count)
+            recs = await self.item_recommender.related(resolved_id, 20)
```
Comment (eu9ene) [inline app/data_providers/dispatch.py:61]: it is not guaranteed that there will be 5 different domains in those 20 items, but let's start with this and increase it later if we still see a lot of duplicates after testing

File: app/data_providers/dispatch.py [hunk 2]
```diff
@@ -78,13 +85,26 @@ async def by_publisher(self, resolved_id: int, domain: str, count: int) -> List[
...
+            else:
+                duplicates.append(r)
+        return uniques + duplicates
 
     @staticmethod
-    def _sample(recs, count):
-        """ Randomly sample articles or apply thompson sampling """
-        recs = random.sample(recs, count) if count < len(recs) else recs
-        return [CorpusRecommendationModel(corpus_item=r) for r in recs]
+    def _to_corpus_items(recs, count):
```
Comment (mmiermans) [inline app/data_providers/dispatch.py:105]: Nice rename. Much clearer. :+1:

File: app/data_providers/dispatch.py [hunk 3]
```diff
@@ -78,13 +85,26 @@ async def by_publisher(self, resolved_id: int, domain: str, count: int) -> List[
...
+        seen_domains = set()
+        duplicates = []
+        uniques = []
+        for r in recs:
+            if r.domain not in seen_domains:
+                uniques.append(r)
+                seen_domains.add(r.domain)
+            else:
+                duplicates.append(r)
+        return uniques + duplicates
```
Comment (mmiermans) [inline app/data_providers/dispatch.py:102]: Similar logic is already implemented in `spread_publishers` and `spread_topics`. Might be good to standardize one a single approach to spread based on a generic attribute.

For this PR, can you move this function to `app/rankers/algorithms.py` so it's easy to keep track of the rankers that we have?

**Review by mmiermans — APPROVED** [review]
I only have some minor suggestions.

If you have a Jira ticket tracking this work, could you link it in the PR description and title?

File: app/data_providers/dispatch.py (hunk 3 above)
Comment (eu9ene) [reply app/data_providers/dispatch.py:102]: I agree, I thought about this, but the requirements, use cases and input/output data structures are quite different with the home pub spread. I can move it to algorithms, yes

**Pull request merged**